### ISSUE TITLE
What should be allowable characters in array names?

### ISSUE THREAD

**Issue by nsheff** [ga4gh/seqcol-spec#33]
The sequence collection is a group of named arrays. These array names include built-in, defined arrays, like `names`, `lengths`, and `sequences`, but users may also use custom array names. Our spec-defined array names are all lowercase ascii characters, but this doesn't mean we must restrict custom array names in the same way. 

Should custom array names be restricted in some way? We want to balance 3 desirable properties: 1) interoperable; 2) flexible; 3) easy-to-implement, but can't come up with a solution that excels at all 3 of these. We have come up with 3 options, each one prioritizing two of the desirable properties:

## Option 1
Make it *required* that custom array names be ascii characters (maybe even lowercase ascii characters).
✔️ Interoperable (because all servers will use the same standard)
✔️ Easy to implement (no questions about UTF sorting or encoding, can use array names easily in API endpoint or database names; specification is simpler)
❌ LESS flexible (Users can't use non-English array names)

## Option 2
Make it *recommended* that custom array names be ascii characters, but allow users to use UTF-8 characters as an extension to the spec. Implementing UTF-8 will not be required for an implementation.
❌ LESS interoperable (servers that don't implement UTF-8 will not be completely compatible with any that do)
✔️ Easy to implement (you don't have to implement UTF-8 unless you need it)
✔️ Flexible (users who want non-English array names can use them)

## Option 3
Make *UTF-8* the specification, so it's required to implement support for UTF-8 array names.
✔️ Interoperable (all servers will be required to allow UTF-8 array names)
❌ LESS easy to implement (implementations **must** consider sorting algorithm, array name normalization, and can't use array names directly as URL endpoints, database identifiers; specification is more complex)
✔️ Flexible (anyone can use non-English array names)

Feedback or corrections appreciated!

**Comment by daviesrob** [comment]
If allowing UTF-8 for array names, I think we should specify that they are converted to the NFKC [normalization form](https://www.unicode.org/reports/tr15/#Norm_Forms).  The Unicode [normalization FAQ](https://www.unicode.org/faq/normalization.html) recommends this for identifiers because it removes several formatting distinctions that could otherwise lead to confusion.  Also worth considering are [rfc3454 - "stringprep"](https://www.ietf.org/rfc/rfc3454.txt) and [rfc3491 - "nameprep"](https://www.ietf.org/rfc/rfc3491.txt) which build on NFKC with additional rules to deal with complications like invisible characters and the wide variety of spaces.  rfc3491 is part of the process used to normalize international names for DNS.

**Comment by sveinugu** [comment]
Some comments from me (and sorry for not attending the meeting yesterday):

First, I wanted to clarify my main point from the last meeting I attended: I strongly believe that saying that we only support ASCII now is really a limitation that will make future support for UTF-8 in the standard a breaking change in relation to the implementations. 

Here is an example of consequences from a previous decision to limit a standard to only support ASCII characters. This is relating to a format/standard named "Exif", which seems to be a standard for metadata of photos: https://exiftool.org/gui/articles/where_what.html:

> "1st limitation: Officially, Exif supports ASCII characters only. However, Metadata Working Group (MWG) organisation recommends using UTF-8 charset in Exif. No problem with that, as long characters 32..126 are used. But as soon we use some foreign characters, then again, there's no guarantee, that these characters will be displayed correctly if another software is used -software might "expect" that characters conforms Exif specification (are ASCII only) and will thus decode foreign characters wrong!"

I believe that just adding the word "UTF-8" to the standard is a way of future-proofing it. It does not matter much that we now mostly (or only) imagine use cases with only ASCII characters. I believe it is highly likely that a use case that requires UTF-8 will arise at some point, and just specifying 'UTF-8' now is a cheap way to insure us against issues later. It's quite a bit like the Y2K situation, really.

> The 3 issues we've identified with allowing UTF-8 in array names are:
> 
> 1. Normalization. The UTF-8 has different ways to represent the same (or very similar) thing, which can lead to identifiability problems if users are not careful. We would probably need to define in the specification some normalization scheme to make sure things a user expects to be identical will hash to the same digest.

I think we can very well say that we support UTF-8 without specifying a normalization scheme at this time. This is especially true on the basis that we do not envision that users are currently up with use cases that require UTF-8 (in which case we should definitely have supported it). Since few users, if any, will likely use the non-ASCII UTF-8 characters, the practical need for a normalization strategy now is low. However, assuming we specify UTF-8 support now, it will be relatively easy to add a future limitations in this direction if the need arises without breaking core functionality in the implementations (although a few seqcols might need to be regenerated, at least in theory).

Of these three choices:

a) Only supporting ASCII
b) Supporting UTF-8
c) Supporting UTF-8 with a particular normalization strategy

I would argue that b) is the most open one, while both a) and c) pose limitations. Only supporting ASCII is a far too conservative limitation, IMO. Specifying c) now might possibly be a gamble in case the rest of the world converge around a different normalization strategy later. Deciding on b) is a no-risk choice, as the rest of the world has already converged around UTF-8.

> 2. Sort order. These arrays are sorted alphabetically the way we defined the built-in arrays, which is simple and synonomous with a byte-code based sorting for ASCII. If we go to other more comprehensive character sets, we'd need to define the sort order. Defining it as byte-based sorting may solve this, but that's then not going to be the same as a natural sort (in case that matters to someone).

Using the UTF-8 byte order for sorting is the natural choice, IMO. I believe nobody assumes anything related to the semantics of sorting in the context of the seqcol internals anyway. Also, the issue is really more or less the same also for ASCII, as we have lived with the lexical sorting of human chromosomes as "chr1, chr10, chr11, ..., chr2, chr20, ..." for twenty years.

The GA4GH Variation Representation Specification group has already handled the details of JSON/digest serialization of UTF-8 encoded strings, including sorting, so I suggest we just copy that:

https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#digest-serialization

> 3. use of array names in other places will be restricted if UTF-8 is allowed. While not part of the current spec, an implementation may want to create API endpoints that use array names. For example, an endpoint to grab just a specific array from a digest could be useful. Or, the array names could be used as table names or in columns in a database. If array names can be UTF-8, it may preclude using them directly in some of these situations, which increases implementation complexity and may make some things impossible.

So for array names, I agree that it would be useful if we limit these to ASCII-only. In addition than what is mentioned above, I have another argument/use case in favor of ASCII array names.

In implementations or downstream scripts, one might want to use the array names as variables in a programming language. Even though most programming languages these days also allow the use of UTF-8 in the programming code itself, the commonly accepted standard is still to code in ASCII only, at least in the parts of the code that are not string constants. So I think it is perfectly fine to restrict array names to be ASCII, with the further constraint of only allowing lower-case characters. I would actually also add to this requirement that array names with multiple words should be styled using `snake_case`.

So in brief, I suggest we restrict the standard to ASCII only, `snake_case` array names (which I believe per definition also entails all lowercase characters). Array values are encoded as UTF-8, following the GA4GH Variation Representation Specification.

Case closed! (in my dreams...)

**Comment by sveinugu** [comment]
> @sveinugu Thanks! But there's one thing I wanted to clarify. You say two conflicting things:
> 
> > Only supporting ASCII is a far too conservative limitation, IMO.

> Just to be totally clear here -- It seems like you're making a distinction between array names and array contents, but to me, the conversation about restricting to ASCII has only ever been about array names. The point of this issue is array names, not contents. I don't think anybody ever suggested restricting array contents to ASCII.

Whoops! Sorry for the misunderstanding. For some reason, my brain read "array names" as "arrays", since it already understood what this was all about...

So I suppose you had a nice and productive meeting without my misunderstanding of the issue botching up everything! 😄 

I suppose I came to my misunderstanding as this topic way raised as a subtopic of the main issue of specifying the digest algorithm. If UTF-8 is allowed in array contents, then the digest algorithm necessarily needs to support UTF-8, and thus its implementations. Hence, my main argument for saying that ASCII is too restrictive falls, as the digest algorithm implementations will anyway support UTF-8. Any implementation issues will then be in the data storage (e.g. database), URL handling, programming language support and similar, and here I agree that specifying that the implementations need to support UTF-8 might limit them technology-wise. Also, any requests for UTF-8 array names will probably come from people that want to extend the standard and thus probably would want to make their own implementations anyway. Lastly, being able to write e.g. ```colseq.my_array``` in a programming language seems very useful (my additional argument).

I suppose we can now really say "case closed" then!

**Comment by sveinugu** [comment]
I think I'm in the option 2 category as well, allowing for UTF-8 but recommending ASCII. In case implementations for some reason are not able to support UTF-8 for array names, at least they will follow the recommendations, while anyone that may want to use UTF-8 array names will do so at their own risk. However, it will probably still work.

**Comment by sveinugu** [comment]
With a proper understanding of the issue, I now have comments to the listed negatives of Option 3:

> ❌ LESS easy to implement (implementations **must** consider sorting algorithm, array name normalization, and can't use array names directly as URL endpoints, database identifiers; specification is more complex)

# Sorting algorithm
The implementers of the **digest algorithm** must consider a sorting algorithm, but if we follow the [GA4GH Variation Representation Specification](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#digest-serialization) this is already clearly defined and should be no big deal.

# Normalization
Array name normalization is in any case mostly an issue for the submitters and not the developers. Also, I don't expect this to become an issue for array names at all. If someone has extended the spec with UTF-8 array names, they will also most likely have produced a machine-operable JSON schema with the exact unicode encoding of the name, which will be the canonical encoding anyway.

#URL endpoints
All browsers and general libraries support IRI's these days anyway (UTF-8-encoded URIs, of which URLs are a subset). I would think most API libraries also support UTF-8 encoded endpoints also these days, but I don't know this. 

#Database identifiers
I would believe those can be UTF-8 encoded in most RDBMS these days, but I am not a database expert. However, some legacy systems might fail in this regard.

**Comment by sveinugu** [comment]
Also, having read the options in more detail, I now see that I also slightly misunderstood option 2:

> ## Option 2
> Make it _recommended_ that custom array names be ascii characters, but allow users to use UTF-8 characters as an extension to the spec. Implementing UTF-8 will not be required for an implementation. ❌ LESS interoperable (servers that don't implement UTF-8 will not be completely compatible with any that do) ✔️ Easy to implement (you don't have to implement UTF-8 unless you need it) ✔️ Flexible (users who want non-English array names can use them)

My option 2 was really a variant of option 2

## Option 2B
Make it _recommended_ that custom array names be ascii characters, but allow users to use UTF-8 characters as an extension to the spec. Implementing UTF-8 will not be required for an implementation, __but recommended__.

_(text in bold type has been added)_

### Reasoning
Basically we would then recommend to users that they restrain themselves to ASCII for array names, while we recommend to implementers that they still support UTF-8, to be more robust for extensions. However, if there are legacy systems that are unable to support UTF-8, they can still rightfully claim to be seqcol compliant. With such a compromise, I don't think we will leave anyone out!

**Comment by nsheff** [comment]
Discussion in December 2023 seemed to lean toward: array names SHOULD be ASCII, but it's up to implementers if they want to stray from that because it's a SHOULD not a MUST

**Comment by nsheff** [comment]
Conclusion from today's meeting:

Option 2, recommend ASCII.
